stackslib/src/chainstate/stacks/boot/pox-4.clar

```diff
@@ -704,21 +734,32 @@
 ;; See `get-signer-key-message-hash` for details on the message hash.
 ;;
 ;; Note that `reward-cycle` corresponds to the _current_ reward cycle,
-;; not the reward cycle at which the delegation will start.
-;; The public key is recovered from the signature and compared to `signer-key`.
+;; when used with `stack-stx` and `stack-extend`.
+;; When `signer-sig` is present, the public key is recovered from the signature 
+;; and compared to `signer-key`.
+;; 
+;; If `signer-sig` is `none`, the function verifies that an authorization was previously
+;; added for this key.
 (define-read-only (verify-signer-key-sig (pox-addr { version: (buff 1), hashbytes: (buff 32) })
                                          (reward-cycle uint)
                                          (topic (string-ascii 12))
                                          (period uint)
-                                         (signer-sig (buff 65))
+                                         (signer-sig-opt (optional (buff 65)))
                                          (signer-key (buff 33)))
-  (ok (asserts! 
-    (is-eq 
-      (unwrap! (secp256k1-recover? 
-        (get-signer-key-message-hash pox-addr reward-cycle topic period) 
-        signer-sig) (err ERR_INVALID_SIGNATURE_RECOVER)) 
-      signer-key) 
-    (err ERR_INVALID_SIGNATURE_PUBKEY))))
+  (match signer-sig-opt
+    ;; `signer-sig` is present, verify the signature
+    signer-sig (ok (asserts! 
+      (is-eq 
+        (unwrap! (secp256k1-recover? 
+          (get-signer-key-message-hash pox-addr reward-cycle topic period) 
+          signer-sig) (err ERR_INVALID_SIGNATURE_RECOVER)) 
+        signer-key) 
+      (err ERR_INVALID_SIGNATURE_PUBKEY)))
+    ;; `signer-sig` is not present, verify that an authorization was previously added for this key
```

Comment:
Based on the way this function is getting used, a couple things stand out to me here that should be documented.

First, **the reward cycle and duration are inflexible.**  The signer and stacker have to agree on _exactly_ the reward cycle _and_ period length before the stacker can stack, or this fails.  For example, if Alice wants signer Sally to be her signer, and Alice initially wanted to stack for 12 reward cycles but changes her mind and stacks for only 6 cycles, this function will fail unless Sally updates the signer record to indicate 6 cycles.  As another example, suppose Sally agrees to sign for Alice from cycle X through X+12, but Alice stacks in cycle X+1 but for only 11 reward cycles, then her stacking transaction fails unless Sally updates her signer key authorization for Alice.  This isn't a bug per se, but it might be surprising to users.

Second, **the caller must validate `reward-cycle` and `period`.**  By itself, this function would accept reward cycles in the past, or period lengths of 0 cycles or more than 12 cycles.  It may make sense to add defensive checks here, but I don't think it's a problem as long as all of the callers perform these checks.  However, I think this needs to be documented.